File: test/language/array_types/cpp98/FixedArrayTest.cpp
#include "gtest/gtest.h"

#include "array_types/fixed_array/FixedArray.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

namespace array_types
{
namespace fixed_array
{

class FixedArrayTest : public ::testing::Test
{
protected:
    void writeFixedArrayToByteArray(zserio::BitStreamWriter& writer)
    {
        for (size_t i = 0; i < FIXED_ARRAY_LENGTH; ++i)
            writer.writeBits(static_cast<uint32_t>(i), 8);
    }

    static const size_t FIXED_ARRAY_LENGTH = 5;
};

TEST_F(FixedArrayTest, bitSizeOf)
{
    zserio::UInt8Array uint8Array;
    uint8Array.reserve(FIXED_ARRAY_LENGTH);
    for (size_t i = 0; i < FIXED_ARRAY_LENGTH; ++i)
        uint8Array.push_back(static_cast<uint8_t>(i));
    FixedArray fixedArray;
    fixedArray.setUint8Array(uint8Array);

    const size_t bitPosition = 2;
    ASSERT_EQ(FIXED_ARRAY_LENGTH * 8, fixedArray.bitSizeOf(bitPosition));
}

TEST_F(FixedArrayTest, initializeOffsets)
{
    zserio::UInt8Array uint8Array;
    uint8Array.reserve(FIXED_ARRAY_LENGTH);
    for (size_t i = 0; i < FIXED_ARRAY_LENGTH; ++i)
        uint8Array.push_back(static_cast<uint8_t>(i));
    FixedArray fixedArray;
    fixedArray.setUint8Array(uint8Array);

    const size_t bitPosition = 2;
    ASSERT_EQ(bitPosition + FIXED_ARRAY_LENGTH * 8, fixedArray.initializeOffsets(bitPosition));
}

TEST_F(FixedArrayTest, read)
{
    zserio::BitStreamWriter writer;
    writeFixedArrayToByteArray(writer);
    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
    FixedArray fixedArray(reader);

    const zserio::UInt8Array& uint8Array = fixedArray.getUint8Array();
    const size_t numElements = FIXED_ARRAY_LENGTH;
    ASSERT_EQ(numElements, uint8Array.size());
    for (size_t i = 0; i < numElements; ++i)
        ASSERT_EQ(i, uint8Array[i]);
}

TEST_F(FixedArrayTest, write)
{
    zserio::UInt8Array uint8Array;
    uint8Array.reserve(FIXED_ARRAY_LENGTH);
    for (size_t i = 0; i < FIXED_ARRAY_LENGTH; ++i)
        uint8Array.push_back(static_cast<uint8_t>(i));
    FixedArray fixedArray;
    fixedArray.setUint8Array(uint8Array);

    zserio::BitStreamWriter writer;
    fixedArray.write(writer);

    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
    FixedArray readFixedArray(reader);
    const zserio::UInt8Array& readUint8Array = readFixedArray.getUint8Array();
    const size_t numElements = FIXED_ARRAY_LENGTH;
    ASSERT_EQ(numElements, readUint8Array.size());
    for (size_t i = 0; i < numElements; ++i)
        ASSERT_EQ(i, readUint8Array[i]);
}

TEST_F(FixedArrayTest, writeWrongArray)
{
    zserio::UInt8Array uint8Array;
    const size_t wrongArrayLength = FIXED_ARRAY_LENGTH + 1;
    uint8Array.reserve(wrongArrayLength);
    for (size_t i = 0; i < wrongArrayLength; ++i)
        uint8Array.push_back(static_cast<uint8_t>(i));
    FixedArray fixedArray;
    fixedArray.setUint8Array(uint8Array);

    zserio::BitStreamWriter writer;
    ASSERT_THROW(fixedArray.write(writer), zserio::CppRuntimeException);
}

} // namespace fixed_array
} // namespace array_types

File: test/language/array_types/python/FixedArrayTest.py
import unittest
import zserio

from testutils import getZserioApi

class FixedArrayTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "array_types.zs").fixed_array

    def testBitSizeOf(self):
        uint8Array = list(range(self.FIXED_ARRAY_LENGTH))
        fixedArray = self.api.FixedArray(uint8Array)
        bitPosition = 2
        fixedArrayBitSize = self.FIXED_ARRAY_LENGTH * 8
        self.assertEqual(fixedArrayBitSize, fixedArray.bitsizeof(bitPosition))

    def testInitializeOffsets(self):
        uint8Array = list(range(self.FIXED_ARRAY_LENGTH))
        fixedArray = self.api.FixedArray(uint8Array)
        bitPosition = 2
        expectedEndBitPosition = bitPosition + self.FIXED_ARRAY_LENGTH * 8
        self.assertEqual(expectedEndBitPosition, fixedArray.initialize_offsets(bitPosition))

    def testRead(self):
        writer = zserio.BitStreamWriter()
        self._writeFixedArrayToStream(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        fixedArray = self.api.FixedArray.from_reader(reader)

        uint8Array = fixedArray.uint8_array
        self.assertEqual(self.FIXED_ARRAY_LENGTH, len(uint8Array))
        for i in range(self.FIXED_ARRAY_LENGTH):
            self.assertEqual(i, uint8Array[i])

    def testWrite(self):
        uint8Array = list(range(self.FIXED_ARRAY_LENGTH))
        fixedArray = self.api.FixedArray(uint8Array)
        bitBuffer = zserio.serialize(fixedArray)
        readFixedArray = zserio.deserialize(self.api.FixedArray, bitBuffer)
        readUint8Array = readFixedArray.uint8_array
        self.assertEqual(self.FIXED_ARRAY_LENGTH, len(readUint8Array))
        for i in range(self.FIXED_ARRAY_LENGTH):
            self.assertEqual(i, readUint8Array[i])

    def testWriteWrongArray(self):
        uint8Array = list(range(self.FIXED_ARRAY_LENGTH + 1))
        fixedArray = self.api.FixedArray(uint8_array_=uint8Array)
        writer = zserio.BitStreamWriter()
        with self.assertRaises(zserio.PythonRuntimeException):
            fixedArray.write(writer)

    def _writeFixedArrayToStream(self, writer):
        for i in range(self.FIXED_ARRAY_LENGTH):
            writer.write_bits(i, 8)

    FIXED_ARRAY_LENGTH = 5
